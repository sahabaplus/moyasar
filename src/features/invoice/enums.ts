export const InvoiceStatus = {
  INITIATED: "initiated",
  PAID: "paid",
  FAILED: "failed",
  REFUNDED: "refunded",
  CANCELED: "canceled",
  ON_HOLD: "on_hold",
  EXPIRED: "expired",
  VOIDED: "voided",
} as const;

export type InvoiceStatus = (typeof InvoiceStatus)[keyof typeof InvoiceStatus];

export const ALL_INVOICE_STATUSES = Object.values(InvoiceStatus);
