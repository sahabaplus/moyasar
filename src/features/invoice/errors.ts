import { MoyasarError } from "@errors";
export class InvoiceError extends MoyasarError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, "INVOICE_ERROR", 500, details ?? {});
    this.name = "InvoiceError";
  }
}
