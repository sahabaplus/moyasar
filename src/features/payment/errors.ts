import { MoyasarError } from "@errors";

export class PaymentError extends MoyasarError {
  constructor(
    message: string,
    statusCode: number,
    details?: Record<string, any>
  ) {
    super(message, "PAYMENT_ERROR", statusCode, details ?? {});
    this.name = "PaymentError";
  }
}
