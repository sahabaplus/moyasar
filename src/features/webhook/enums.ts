export const WebhookEvents = {
  PAYMENT_PAID: "payment_paid",
  PAYMENT_FAILED: "payment_failed",
  PAYMENT_VOIDED: "payment_voided",
  PAYMENT_AUTHORIZED: "payment_authorized",
  PAYMENT_CAPTURED: "payment_captured",
  PAYMENT_REFUNDED: "payment_refunded",
  PAYMENT_ABANDONED: "payment_abandoned",
  PAYMENT_VERIFIED: "payment_verified",
  PAYMENT_CANCELED: "payment_canceled",
  PAYMENT_EXPIRED: "payment_expired",
  BALANCE_TRANSFERRED: "balance_transferred",
  PAYOUT_INITIATED: "payout_initiated",
  PAYOUT_PAID: "payout_paid",
  PAYOUT_FAILED: "payout_failed",
  PAYOUT_CANCELED: "payout_canceled",
  PAYOUT_RETURNED: "payout_returned",
} as const;

export type WebhookEvents = (typeof WebhookEvents)[keyof typeof WebhookEvents];

export const ALL_WEBHOOK_EVENTS = Object.values(WebhookEvents);

export const WebhookHttpMethods = {
  POST: "post",
  PUT: "put",
  PATCH: "patch",
} as const;
