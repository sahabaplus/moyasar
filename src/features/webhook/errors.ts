import { MoyasarError } from "@errors";
import type { WebhookPayload } from "./types";

export class WebhookError extends MoyasarError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, "WEBHOOK_ERROR", 500, details ?? {});
    this.name = "WebhookError";
  }
}

export class WebhookVerificationError extends WebhookError {
  constructor(message: string = "Webhook signature verification failed") {
    super(message, {});
    this.name = "WebhookVerificationError";
  }
}

export class WebhookValidationError extends WebhookError {
  public readonly unexpected_payload: Partial<WebhookPayload>;
  constructor({
    message = "Webhook payload validation failed",
    unexpected_payload,
  }: {
    message?: string;
    unexpected_payload: Partial<WebhookPayload>;
  }) {
    super(message, { unexpected_payload });
    this.unexpected_payload = unexpected_payload;
    this.name = "WebhookValidationError";
  }
}
