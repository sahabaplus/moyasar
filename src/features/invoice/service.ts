import type { ApiClient, MoyasarClientTypes } from "@types";
import { API_ENDPOINTS } from "@constants";
import { MoyasarError } from "@errors";
import type {
  CreateInvoiceRequest,
  UpdateInvoiceRequest,
  BulkCreateInvoiceRequest,
  ListInvoicesResponse,
  BulkCreateInvoicesResponse,
  InvoiceListOptions,
  DetailedInvoice,
} from "./types";
import { InvoiceStatus } from "./enums";
import { InvoiceUtils } from "./utils";
import { InvoiceError } from "./errors";

type InvoiceServiceParams<T extends MoyasarClientTypes> = {
  apiClient: ApiClient<T>;
};

export class InvoiceService<T extends MoyasarClientTypes> {
  private readonly apiClient: ApiClient<T>;
  private readonly invoiceUtils: InvoiceUtils<T["metadata"]>;

  constructor(p: InvoiceServiceParams<T>) {
    this.apiClient = p.apiClient;
    this.invoiceUtils = new InvoiceUtils({
      metadataValidator: this.apiClient.metadataValidator,
    });
  }

  /**
   * Create a new invoice
   */
  async create(
    params: CreateInvoiceRequest<T["metadata"]>
  ): Promise<DetailedInvoice<T["metadata"]>> {
    // Validate input
    const validation = this.invoiceUtils.validateCreateInvoiceRequest(params);
    if (!validation.success) {
      throw new InvoiceError(
        `Validation failed: ${validation.errors.join(", ")}`
      );
    }

    try {
      const invoice = await this.apiClient.request<
        DetailedInvoice<T["metadata"]>
      >({
        method: "POST",
        url: API_ENDPOINTS.invoices,
        data: params,
      });

      return invoice;
    } catch (error) {
      const invoiceError = this.handleError(error, "Failed to create invoice");
      throw invoiceError;
    }
  }

  /**
   * Create multiple invoices in bulk
   */
  async createBulk(
    params: BulkCreateInvoiceRequest<T["metadata"]>
  ): Promise<BulkCreateInvoicesResponse<T["metadata"]>> {
    // Validate input
    const validation = this.invoiceUtils.validateBulkCreateRequest(params);
    if (!validation.success) {
      throw new InvoiceError(
        `Validation failed: ${validation.errors.join(", ")}`
      );
    }

    try {
      const response = await this.apiClient.request<
        BulkCreateInvoicesResponse<T["metadata"]>
      >({
        method: "POST",
        url: API_ENDPOINTS.bulkInvoices,
        data: params,
      });

      return response;
    } catch (error) {
      const invoiceError = this.handleError(
        error,
        "Failed to create bulk invoices"
      );
      throw invoiceError;
    }
  }

  /**
   * List invoices with optional filtering
   */
  async list(
    options: InvoiceListOptions<T["metadata"]> = {}
  ): Promise<ListInvoicesResponse<T["metadata"]>> {
    try {
      // Convert metadata filters to proper query format
      const queryParams = this.parseBody(options);

      return await this.apiClient.request<ListInvoicesResponse<T["metadata"]>>({
        method: "GET",
        url: API_ENDPOINTS.invoices,
        params: queryParams,
      });
    } catch (error) {
      const invoiceError = this.handleError(error, "Failed to list invoices");
      throw invoiceError;
    }
  }

  /**
   * Retrieve a specific invoice
   */
  async retrieve(invoiceId: string): Promise<DetailedInvoice<T["metadata"]>> {
    if (!invoiceId) {
      throw new InvoiceError("Invoice ID is required");
    }

    try {
      return await this.apiClient.request<DetailedInvoice<T["metadata"]>>({
        method: "GET",
        url: `${API_ENDPOINTS.invoices}/${invoiceId}`,
      });
    } catch (error) {
      const invoiceError = this.handleError(
        error,
        `Failed to retrieve invoice ${invoiceId}`
      );
      throw invoiceError;
    }
  }

  /**
   * Update an invoice
   */
  async update(
    invoiceId: string,
    params: UpdateInvoiceRequest<T["metadata"]>
  ): Promise<DetailedInvoice<T["metadata"]>> {
    if (!invoiceId) {
      throw new InvoiceError("Invoice ID is required");
    }

    try {
      const invoice = await this.apiClient.request<
        DetailedInvoice<T["metadata"]>
      >({
        method: "PUT",
        url: `${API_ENDPOINTS.invoices}/${invoiceId}`,
        data: params,
      });

      return invoice;
    } catch (error) {
      const invoiceError = this.handleError(
        error,
        `Failed to update invoice ${invoiceId}`
      );
      throw invoiceError;
    }
  }

  /**
   * Cancel an invoice
   */
  async cancel(invoiceId: string): Promise<DetailedInvoice<T["metadata"]>> {
    if (!invoiceId) {
      throw new InvoiceError("Invoice ID is required");
    }

    try {
      const invoice = await this.apiClient.request<
        DetailedInvoice<T["metadata"]>
      >({
        method: "PUT",
        url: `${API_ENDPOINTS.invoices}/${invoiceId}/cancel`,
      });

      return invoice;
    } catch (error) {
      const invoiceError = this.handleError(
        error,
        `Failed to cancel invoice ${invoiceId}`
      );
      throw invoiceError;
    }
  }

  /**
   * Search invoices by metadata
   */
  async searchByMetadata(
    metadata: Partial<T["metadata"]>,
    options: Omit<InvoiceListOptions, "metadata"> = {}
  ): Promise<ListInvoicesResponse<T["metadata"]>> {
    return this.list({
      ...options,
      metadata,
    });
  }

  /**
   * Get invoices by status
   */
  async getByStatus(
    status: InvoiceStatus,
    options: Omit<InvoiceListOptions, "status"> = {}
  ): Promise<ListInvoicesResponse<T["metadata"]>> {
    return this.list({
      ...options,
      status,
    });
  }

  /**
   * Get expired invoices
   */
  async getExpired(
    options: Omit<InvoiceListOptions, "status"> = {}
  ): Promise<ListInvoicesResponse<T["metadata"]>> {
    return this.getByStatus(InvoiceStatus.EXPIRED, options);
  }

  /**
   * Get paid invoices
   */
  async getPaid(
    options: Omit<InvoiceListOptions, "status"> = {}
  ): Promise<ListInvoicesResponse<T["metadata"]>> {
    return this.getByStatus(InvoiceStatus.PAID, options);
  }

  private handleError(error: any, message: string): InvoiceError {
    if (error instanceof InvoiceError) {
      return error;
    }

    if (error instanceof MoyasarError) {
      return new InvoiceError(`${message}: ${error.message}`, {
        ...error.details,
      });
    }

    const errorMessage = error?.message || error?.toString() || "Unknown error";
    return new InvoiceError(`${message}: ${errorMessage}`, {
      cause: errorMessage,
    });
  }

  protected parseBody(p: InvoiceListOptions<T["metadata"]>) {
    const copied = {
      ...p,
      ...this.invoiceUtils.buildMetadataQuery(p.metadata ?? {}),
    };

    delete copied.metadata;

    Object.entries(copied).forEach(([key, value]) => {
      if (typeof value === "object") {
        if (value instanceof Date) {
          // @ts-expect-error
          copied[key] = value.toISOString();
        }
      }
    });

    return copied;
  }
}
