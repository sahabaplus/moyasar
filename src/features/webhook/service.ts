import type { ApiClient, Metadata, MoyasarClientTypes } from "@types";
import { API_ENDPOINTS } from "@constants";
import { WebhookUtils } from "./utils";
import { WebhookError, WebhookValidationError } from "./errors";
import { WebhookValidation } from "./validation";
import { WebhookEvent } from "./enums";
import type {
  Webhook,
  WebhookPayload,
  WebhookAttempt,
  CreateWebhookRequest,
  UpdateWebhookRequest,
  ListWebhooksResponse,
  ListWebhookAttemptsResponse,
  AvailableEventsResponse,
  WebhookListOptions,
  WebhookAttemptListOptions,
  WebhookVerificationOptions,
  WebhookEventMap,
} from "./types";
import { TypedEmitter } from "tiny-typed-emitter";

type WebhookServiceParams<T extends MoyasarClientTypes> = {
  apiClient: ApiClient<T>;
};

export class WebhookService<T extends MoyasarClientTypes> extends TypedEmitter<
  WebhookEventMap<T["metadata"]>
> {
  private readonly apiClient: ApiClient<T>;
  private readonly events = Object.values(WebhookEvent) as WebhookEvent[];
  constructor(params: WebhookServiceParams<T>) {
    super();
    this.apiClient = params.apiClient;
  }

  /**
   * Create a new webhook
   */
  async create(params: CreateWebhookRequest): Promise<Webhook> {
    // Validate input
    this.validateCreateRequest(params);

    try {
      const webhook = await this.apiClient.request<Webhook>({
        method: "POST",
        url: API_ENDPOINTS.webhooks,
        data: params,
      });

      return webhook;
    } catch (error) {
      const webhookError = this.handleError(error, "Failed to create webhook");
      throw webhookError;
    }
  }

  /**
   * List all webhooks
   */
  async list(options: WebhookListOptions = {}): Promise<ListWebhooksResponse> {
    try {
      return await this.apiClient.request<ListWebhooksResponse>({
        method: "GET",
        url: API_ENDPOINTS.webhooks,
        params: options,
      });
    } catch (error) {
      const webhookError = this.handleError(error, "Failed to list webhooks");
      throw webhookError;
    }
  }

  /**
   * Retrieve a specific webhook
   */
  async retrieve(webhookId: string): Promise<Webhook> {
    if (!webhookId) {
      throw new WebhookError("Webhook ID is required");
    }

    try {
      return await this.apiClient.request<Webhook>({
        method: "GET",
        url: `${API_ENDPOINTS.webhooks}/${webhookId}`,
      });
    } catch (error) {
      const webhookError = this.handleError(
        error,
        `Failed to retrieve webhook ${webhookId}`
      );
      throw webhookError;
    }
  }

  /**
   * Update a webhook
   */
  async update(
    webhookId: string,
    params: UpdateWebhookRequest
  ): Promise<Webhook> {
    if (!webhookId) {
      throw new WebhookError("Webhook ID is required");
    }

    // Validate input
    this.validateUpdateRequest(params);

    try {
      const webhook = await this.apiClient.request<Webhook>({
        method: "PUT",
        url: `${API_ENDPOINTS.webhooks}/${webhookId}`,
        data: params,
      });

      return webhook;
    } catch (error) {
      const webhookError = this.handleError(
        error,
        `Failed to update webhook ${webhookId}`
      );
      throw webhookError;
    }
  }

  /**
   * Delete a webhook
   */
  async delete(webhookId: string): Promise<void> {
    if (!webhookId) {
      throw new WebhookError("Webhook ID is required");
    }

    try {
      await this.apiClient.request({
        method: "DELETE",
        url: `${API_ENDPOINTS.webhooks}/${webhookId}`,
      });
    } catch (error) {
      const webhookError = this.handleError(
        error,
        `Failed to delete webhook ${webhookId}`
      );
      throw webhookError;
    }
  }

  /**
   * Get available webhook events
   */
  async availableEvents(): Promise<AvailableEventsResponse> {
    try {
      return await this.apiClient.request<AvailableEventsResponse>({
        method: "GET",
        url: API_ENDPOINTS.availableEvents,
      });
    } catch (error) {
      const webhookError = this.handleError(
        error,
        "Failed to fetch available events"
      );
      throw webhookError;
    }
  }

  /**
   * Webhook Attempts Sub-service
   */
  get attempts() {
    return {
      /**
       * List all webhook attempts
       */
      list: async (
        options: WebhookAttemptListOptions = {}
      ): Promise<ListWebhookAttemptsResponse> => {
        try {
          return await this.apiClient.request<ListWebhookAttemptsResponse>({
            method: "GET",
            url: API_ENDPOINTS.webhookAttempts,
            params: options,
          });
        } catch (error) {
          const webhookError = this.handleError(
            error,
            "Failed to list webhook attempts"
          );
          throw webhookError;
        }
      },

      /**
       * Retrieve a specific webhook attempt
       */
      retrieve: async (attemptId: string): Promise<WebhookAttempt> => {
        if (!attemptId) {
          throw new WebhookError("Attempt ID is required");
        }

        try {
          return await this.apiClient.request<WebhookAttempt>({
            method: "GET",
            url: `${API_ENDPOINTS.webhookAttempts}/${attemptId}`,
          });
        } catch (error) {
          const webhookError = this.handleError(
            error,
            `Failed to retrieve webhook attempt ${attemptId}`
          );
          throw webhookError;
        }
      },
    };
  }

  /**
   * Process incoming webhook payload
   * This method should be called from your webhook endpoint
   */
  async processWebhook(
    rawPayload: string | Buffer | WebhookPayload<Metadata>,
    options: WebhookVerificationOptions
  ): Promise<WebhookPayload<T["metadata"]>> {
    try {
      // Parse payload if it's raw
      const payload =
        typeof rawPayload === "object" && "id" in rawPayload
          ? rawPayload
          : WebhookUtils.parseWebhookPayload(rawPayload as string | Buffer);

      // Validate payload structure
      const validationErrors = WebhookUtils.validateWebhookPayload(payload);
      if (validationErrors.length > 0) {
        throw new WebhookValidationError({
          message: `Invalid webhook payload: ${validationErrors.join(", ")}`,
          unexpected_payload: payload,
        });
      }

      // Verify signature if provided

      const isValidSignature = await WebhookUtils.verifyWebhookSignature(
        payload,
        options
      );
      if (!isValidSignature)
        throw new WebhookError("Webhook signature verification failed");

      try {
        const parsedPayloadData = this.apiClient.metadataValidator.parse(
          payload.data
        );
        const parsedPayload = {
          ...payload,
          data: parsedPayloadData,
        };

        this.emit(payload.type, parsedPayload);
        return parsedPayload;
      } catch (error) {
        throw new WebhookValidationError({
          message:
            error instanceof Error && "message" in error
              ? error.message
              : "Unknown error",
          unexpected_payload: payload,
        });
      }
    } catch (error) {
      const webhookError =
        error instanceof WebhookError
          ? error
          : this.handleError(error, "Failed to process webhook");

      throw webhookError;
    }
  }

  /**
   * Utility method to create type-safe event listeners for payment events
   */
  onPaymentEvent(
    event: WebhookEvent,
    listener: (payload: WebhookPayload<T["metadata"]>) => void | Promise<void>
  ): this {
    return this.on(event, listener);
  }

  /**
   * Utility method to listen to all payment events
   */
  onAnyPaymentEvent(
    listener: (payload: WebhookPayload<T["metadata"]>) => void | Promise<void>
  ): this {
    this.events.forEach(event => {
      this.on(event, listener);
    });

    return this;
  }

  /**
   * Helper method to extract webhook signature from common header patterns
   */
  extractSignature(headers: Record<string, string | string[]>): string | null {
    return WebhookUtils.extractSignatureFromHeaders(headers);
  }

  // Private helper methods

  private validateCreateRequest(params: CreateWebhookRequest): void {
    const errors: string[] = [];

    // Required fields
    if (!params.url) {
      errors.push("url is required");
    } else if (!WebhookValidation.isValidUrl(params.url)) {
      errors.push("url must be a valid HTTP/HTTPS URL");
    }

    if (!params.http_method) {
      errors.push("http_method is required");
    } else if (!WebhookValidation.isValidHttpMethod(params.http_method)) {
      errors.push("http_method must be a valid HTTP method");
    }

    // Validate events if provided
    if (params.events) {
      const invalidEvents = params.events.filter(
        event => !WebhookValidation.isValidWebhookEvent(event)
      );
      if (invalidEvents.length > 0) {
        errors.push(`Invalid webhook events: ${invalidEvents.join(", ")}`);
      }
    }

    if (errors.length > 0) {
      throw new WebhookError(`Validation failed: ${errors.join(", ")}`);
    }
  }

  private validateUpdateRequest(params: UpdateWebhookRequest): void {
    const errors: string[] = [];

    // Validate URL if provided
    if (params.url && !WebhookValidation.isValidUrl(params.url)) {
      errors.push("url must be a valid HTTP/HTTPS URL");
    }

    // Validate HTTP method if provided
    if (
      params.http_method &&
      !WebhookValidation.isValidHttpMethod(params.http_method)
    ) {
      errors.push("http_method must be a valid HTTP method");
    }

    // Validate events if provided
    if (params.events) {
      const invalidEvents = params.events.filter(
        event => !WebhookValidation.isValidWebhookEvent(event)
      );
      if (invalidEvents.length > 0) {
        errors.push(`Invalid webhook events: ${invalidEvents.join(", ")}`);
      }
    }

    if (errors.length > 0) {
      throw new WebhookError(`Validation failed: ${errors.join(", ")}`);
    }
  }

  private handleError(error: any, message: string): WebhookError {
    if (error instanceof WebhookError) {
      return error;
    }

    const errorMessage = error?.message || error?.toString() || "Unknown error";
    return new WebhookError(`${message}: ${errorMessage}`, {
      statusCode: error?.statusCode || error?.status,
      message: errorMessage,
    });
  }
}
