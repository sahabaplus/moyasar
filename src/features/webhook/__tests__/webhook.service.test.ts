import { WebhookEvents, WebhookHttpMethods } from "../enums";
import { WebhookService } from "../service";
import { WebhookError } from "../errors";
import type {
  Webhook,
  WebhookPayload,
  CreateWebhookRequest,
  UpdateWebhookRequest,
  ListWebhooksResponse,
  WebhookAttempt,
} from "../types";
import { MockApiClient } from "./mock_api_client";

describe("WebhookService", () => {
  let mockApiClient: MockApiClient;
  let webhookService: WebhookService;

  beforeEach(() => {
    mockApiClient = new MockApiClient();
    webhookService = new WebhookService({ apiClient: mockApiClient });
  });

  afterEach(() => {
    webhookService.removeAllListeners();
    mockApiClient.clearHistory();
  });

  describe("API tests", async () => {
    it("should accepts the right types and has the right signature", () => {
      const v = webhookService
        .onPaymentEvent(WebhookEvents.PAYMENT_ABANDONED, p => {})
        .onPaymentEvent(WebhookEvents.PAYMENT_ABANDONED, p => {})
        .onAnyPaymentEvent(p => {}).availableEvents;
    });
  });

  describe("create", () => {
    const validCreateRequest: CreateWebhookRequest = {
      url: "https://example.com/webhook",
      http_method: WebhookHttpMethods.POST,
      events: [WebhookEvents.PAYMENT_PAID, WebhookEvents.PAYMENT_FAILED],
      shared_secret: "secret123",
    };

    const mockWebhook: Webhook = {
      id: "webhook_123",
      url: "https://example.com/webhook",
      http_method: WebhookHttpMethods.POST,
      events: [WebhookEvents.PAYMENT_PAID, WebhookEvents.PAYMENT_FAILED],
      created_at: "2024-01-01T00:00:00Z",
      shared_secret: null as never,
    };

    it("should create a webhook successfully", async () => {
      mockApiClient.setMockResponse("POST:/v1/webhooks", mockWebhook);

      const result = await webhookService.create(validCreateRequest);

      expect(result).toEqual(mockWebhook);
      expect(mockApiClient.getLastRequest()).toEqual({
        method: "POST",
        url: "/v1/webhooks",
        data: validCreateRequest,
      });
    });

    it("should validate required fields", async () => {
      const invalidRequest = { ...validCreateRequest, url: "" };

      await expect(webhookService.create(invalidRequest)).rejects.toThrow(
        WebhookError
      );
    });

    it("should validate URL format", async () => {
      const invalidRequest = { ...validCreateRequest, url: "not-a-url" };

      await expect(webhookService.create(invalidRequest)).rejects.toThrow(
        "url must be a valid HTTP/HTTPS URL"
      );
    });

    it("should validate HTTP method", async () => {
      const invalidRequest = {
        ...validCreateRequest,
        http_method: "invalid" as any,
      };

      await expect(webhookService.create(invalidRequest)).rejects.toThrow(
        "http_method must be a valid HTTP method"
      );
    });

    it("should validate event types", async () => {
      const invalidRequest = {
        ...validCreateRequest,
        events: ["invalid_event"] as any,
      };

      await expect(webhookService.create(invalidRequest)).rejects.toThrow(
        "Invalid webhook events: invalid_event"
      );
    });

    it("should handle API errors", async () => {
      const apiError = new Error("API Error");
      mockApiClient.setMockResponse("POST:/v1/webhooks", apiError);

      await expect(webhookService.create(validCreateRequest)).rejects.toThrow();
    });
  });

  describe("list", () => {
    const mockListResponse: ListWebhooksResponse = {
      webhooks: [
        {
          id: "webhook_1",
          url: "https://example.com/webhook1",
          http_method: WebhookHttpMethods.POST,
          events: [WebhookEvents.PAYMENT_PAID],
          created_at: "2024-01-01T00:00:00Z",
          shared_secret: null as never,
        },
      ],
      meta: {
        current_page: 1,
        next_page: null,
        prev_page: null,
        total_pages: 1,
        total_count: 1,
      },
    };

    it("should list webhooks successfully", async () => {
      mockApiClient.setMockResponse("GET:/v1/webhooks", mockListResponse);

      const result = await webhookService.list();

      expect(result).toEqual(mockListResponse);
      expect(mockApiClient.getLastRequest()).toEqual({
        method: "GET",
        url: "/v1/webhooks",
        params: {},
      });
    });

    it("should pass pagination options", async () => {
      mockApiClient.setMockResponse("GET:/v1/webhooks", mockListResponse);

      await webhookService.list({ page: 2, limit: 40 });

      expect(mockApiClient.getLastRequest()!.params).toEqual({
        page: 2,
        limit: 40,
      });
    });
  });

  describe("retrieve", () => {
    const mockWebhook: Webhook = {
      id: "webhook_123",
      url: "https://example.com/webhook",
      http_method: WebhookHttpMethods.POST,
      events: [WebhookEvents.PAYMENT_PAID],
      created_at: "2024-01-01T00:00:00Z",
      shared_secret: null as never,
    };

    it("should retrieve a webhook successfully", async () => {
      mockApiClient.setMockResponse(
        "GET:/v1/webhooks/webhook_123",
        mockWebhook
      );

      const result = await webhookService.retrieve("webhook_123");

      expect(result).toEqual(mockWebhook);
    });

    it("should throw error for empty webhook ID", async () => {
      await expect(webhookService.retrieve("")).rejects.toThrow(
        "Webhook ID is required"
      );
    });
  });

  describe("update", () => {
    const updateRequest: UpdateWebhookRequest = {
      url: "https://updated.example.com/webhook",
      events: [WebhookEvents.PAYMENT_PAID, WebhookEvents.PAYMENT_REFUNDED],
    };

    const mockUpdatedWebhook: Webhook = {
      id: "webhook_123",
      url: "https://updated.example.com/webhook",
      http_method: WebhookHttpMethods.POST,
      events: [WebhookEvents.PAYMENT_PAID, WebhookEvents.PAYMENT_REFUNDED],
      created_at: "2024-01-01T00:00:00Z",
      shared_secret: null as never,
    };

    it("should update a webhook successfully", async () => {
      mockApiClient.setMockResponse(
        "PUT:/v1/webhooks/webhook_123",
        mockUpdatedWebhook
      );

      const result = await webhookService.update("webhook_123", updateRequest);

      expect(result).toEqual(mockUpdatedWebhook);
      expect(mockApiClient.getLastRequest()).toEqual({
        method: "PUT",
        url: "/v1/webhooks/webhook_123",
        data: updateRequest,
      });
    });
  });

  describe("delete", () => {
    it("should delete a webhook successfully", async () => {
      mockApiClient.setMockResponse("DELETE:/v1/webhooks/webhook_123", {});

      await webhookService.delete("webhook_123");

      expect(mockApiClient.getLastRequest()).toEqual({
        method: "DELETE",
        url: "/v1/webhooks/webhook_123",
      });
    });
  });

  describe("processWebhook", () => {
    const validPayload: WebhookPayload = {
      id: "event_123",
      type: WebhookEvents.PAYMENT_PAID,
      created_at: "2024-01-01T00:00:00Z",
      secret_token: "secret",
      account_name: "test_account",
      live: false,
      data: { payment_id: "pay_123" },
    };

    it("should process webhook payload successfully", async () => {
      const eventSpy = jest.fn();
      webhookService.on(WebhookEvents.PAYMENT_PAID, eventSpy);

      const result = await webhookService.processWebhook(validPayload, {
        secret_token: "secret",
      });

      expect(result).toEqual(validPayload);
      expect(eventSpy).toHaveBeenCalledWith(validPayload);
    });

    it("should parse JSON string payload", async () => {
      const jsonPayload = JSON.stringify(validPayload);
      const eventSpy = jest.fn();
      webhookService.on(WebhookEvents.PAYMENT_PAID, eventSpy);

      const result = await webhookService.processWebhook(jsonPayload, {
        secret_token: "secret",
      });

      expect(result).toEqual(validPayload);
      expect(eventSpy).toHaveBeenCalledWith(validPayload);
    });

    it("should validate payload structure", async () => {
      const invalidPayload = { ...validPayload, id: undefined };

      await expect(
        webhookService.processWebhook(invalidPayload as any, {
          secret_token: "secret",
        })
      ).rejects.toThrow(WebhookError);
    });

    it("should reject webhook with wrong secret token", async () => {
      const payloadWithWrongSecret = {
        ...validPayload,
        secret_token: "wrong_secret",
      };

      await expect(
        webhookService.processWebhook(payloadWithWrongSecret, {
          secret_token: "correct_secret",
        })
      ).rejects.toThrow("Webhook signature verification failed");
    });

    it("should reject webhook when secret token is missing in options", async () => {
      await expect(
        webhookService.processWebhook(validPayload, {
          secret_token: "",
        })
      ).rejects.toThrow("Webhook signature verification failed");
    });

    it("should reject webhook when payload secret token is missing", async () => {
      const payloadWithoutSecret = {
        ...validPayload,
        secret_token: "",
      };

      await expect(
        webhookService.processWebhook(payloadWithoutSecret, {
          secret_token: "secret",
        })
      ).rejects.toThrow("Webhook signature verification failed");
    });
  });

  describe("event listeners", () => {
    it("should support onPaymentEvent utility", async () => {
      const paymentSpy = jest.fn();
      webhookService.onPaymentEvent(WebhookEvents.PAYMENT_PAID, paymentSpy);

      const payload: WebhookPayload = {
        id: "event_123",
        type: WebhookEvents.PAYMENT_PAID,
        created_at: "2024-01-01T00:00:00Z",
        secret_token: "secret",
        account_name: "test_account",
        live: false,
        data: { payment_id: "pay_123" },
      };

      await webhookService.processWebhook(payload, {
        secret_token: "secret",
      });

      expect(paymentSpy).toHaveBeenCalledWith(payload);
    });

    it("should support onAnyPaymentEvent utility", async () => {
      const anySpy = jest.fn();
      webhookService.onAnyPaymentEvent(anySpy);

      const payload: WebhookPayload = {
        id: "event_123",
        type: WebhookEvents.PAYMENT_FAILED,
        created_at: "2024-01-01T00:00:00Z",
        secret_token: "secret",
        account_name: "test_account",
        live: false,
        data: { payment_id: "pay_123" },
      };

      await webhookService.processWebhook(payload, {
        secret_token: "secret",
      });

      expect(anySpy).toHaveBeenCalledWith(payload);
    });

    it("should support onWebhookManagement utility", async () => {
      const mockWebhook: Webhook = {
        id: "webhook_123",
        url: "https://example.com/webhook",
        http_method: WebhookHttpMethods.POST,
        events: [WebhookEvents.PAYMENT_PAID],
        created_at: "2024-01-01T00:00:00Z",
        shared_secret: null as never,
      };

      mockApiClient.setMockResponse("POST:/v1/webhooks", mockWebhook);
      mockApiClient.setMockResponse("DELETE:/v1/webhooks/webhook_123", {});

      await webhookService.create({
        url: "https://example.com/webhook",
        http_method: WebhookHttpMethods.POST,
        shared_secret: "secret",
      });

      await webhookService.delete("webhook_123");

      expect(mockApiClient.getLastRequest()).toEqual({
        method: "DELETE",
        url: "/v1/webhooks/webhook_123",
      });
    });
  });

  describe("attempts sub-service", () => {
    const mockAttempt: WebhookAttempt = {
      id: "attempt_123",
      webhook_id: "webhook_123",
      event_id: "event_123",
      event_type: WebhookEvents.PAYMENT_PAID,
      retry_number: 1,
      result: "success",
      message: "Success",
      response_code: 200,
      response_headers: "{}",
      response_body: "OK",
      created_at: "2024-01-01T00:00:00Z",
    };

    it("should list webhook attempts", async () => {
      const mockResponse = {
        webhook_attempts: [mockAttempt],
        meta: {
          current_page: 1,
          next_page: null,
          prev_page: null,
          total_pages: 1,
          total_count: 1,
        },
      };

      mockApiClient.setMockResponse("GET:/v1/webhooks/attempts", mockResponse);

      const result = await webhookService.attempts.list();

      expect(result).toEqual(mockResponse);
    });

    it("should retrieve a specific attempt", async () => {
      mockApiClient.setMockResponse(
        "GET:/v1/webhooks/attempts/attempt_123",
        mockAttempt
      );

      const result = await webhookService.attempts.retrieve("attempt_123");

      expect(result).toEqual(mockAttempt);
    });

    it("should pass filter options to attempts list", async () => {
      mockApiClient.setMockResponse("GET:/v1/webhooks/attempts", {
        webhook_attempts: [],
        meta: {
          current_page: 1,
          next_page: null,
          prev_page: null,
          total_pages: 0,
          total_count: 0,
        },
      });

      await webhookService.attempts.list({
        webhook_id: "webhook_123",
        result: "failed",
        page: 2,
      });

      expect(mockApiClient.getLastRequest()!.params).toEqual({
        webhook_id: "webhook_123",
        result: "failed",
        page: 2,
      });
    });
  });

  describe("extractSignature utility", () => {
    it("should extract signature from common headers", () => {
      const headers = {
        "x-moyasar-signature": "abc123",
        "content-type": "application/json",
      };

      const signature = webhookService.extractSignature(headers);

      expect(signature).toBe("abc123");
    });

    it("should handle signature prefixes", () => {
      const headers = {
        "x-signature": "sha256=abc123",
      };

      const signature = webhookService.extractSignature(headers);

      expect(signature).toBe("abc123");
    });

    it("should return null if no signature found", () => {
      const headers = {
        "content-type": "application/json",
      };

      const signature = webhookService.extractSignature(headers);

      expect(signature).toBeNull();
    });
  });
});
