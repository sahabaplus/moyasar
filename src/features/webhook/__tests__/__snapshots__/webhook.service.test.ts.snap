// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookService Type safe data Using data parser should throw error if payload data is invalid 1`] = `
"[
  {
    "code": "invalid_value",
    "values": [
      "delivery",
      "pickup"
    ],
    "path": [
      "service"
    ],
    "message": "Invalid option: expected one of \\"delivery\\"|\\"pickup\\""
  }
]"
`;

exports[`WebhookService Type safe data Using data parser should throw error as payload data is invalid 1`] = `
"[
  {
    "code": "invalid_value",
    "values": [
      "delivery",
      "pickup"
    ],
    "path": [
      "service"
    ],
    "message": "Invalid option: expected one of \\"delivery\\"|\\"pickup\\""
  }
]"
`;
